File: Sorting_Algos/Heap sort/heap_sort.cpp
#include<bits/stdc++.h>
using namespace std;
void heapsort(int *arr,int s);
void build_max_heap(int *arr,int s);
void max_heapify(int *arr,int i,int s);
void input_array(int *arr,int s);
void print_array(int *arr,int s);
int left(int i);
int right(int i);

int main(void)
{
    int n;
    cout<<"Enter array size: ";
    cin>>n;
    int arr[n];
    input_array(arr,n);
    cout<<"Before sorting"<<endl;
    print_array(arr,n);
    heapsort(arr,n);
    cout<<"After sorting:"<<endl;
    print_array(arr,n);

    return 0;
}
void input_array(int *arr,int s)
{
    for(int i=0;i<s;i++)
        cin>>arr[i];
}
void print_array(int *arr,int s)
{
    for(int i=0;i<s;i++)
        cout<<arr[i]<<" ";
    cout<<endl;
}
void heapsort(int *arr,int s)
{
    build_max_heap(arr,s); /// comes back with the max or min heap
    for(int i= s ;i> 1;i--) ///heap sort here
    {
        swap(arr[0],arr[i - 1]);/// swaps 1st and last element
        s--;
        max_heapify(arr,1,s);
    }

}
void build_max_heap(int *arr,int s)
{
    for(int i = floor(s/2);i>0;i--) /// tree height
        max_heapify(arr,i,s);
}
void max_heapify(int *arr,int i,int s)
{

    int l,r,largest;
    l = left(i);
    r = right(i);
    if(l <= s and arr[l - 1] > arr[i - 1] )
            largest = l;
    else
        largest = i;
    if(r <= s &&  arr[r - 1] > arr[largest - 1] )
            largest = r;

    if(largest != i)
    {
        int temp = arr[i - 1];
        arr[i - 1] = arr[largest - 1];
        arr[largest - 1] = temp;
        max_heapify(arr,largest,s);
    }
}
int left(int i)
{
    return 2*i;

}
int right(int i)
 {
     return (2*i) + 1;
 }

File: Sorting_Algos/Heap sort/heap_sort.py
# -*- coding: utf-8 -*-
"""
Created on Mon Apr 26 20:44:19 2021

@author: User
"""

class HeapSort:
    def __init__(self,my_list,x):
        self.my_list = my_list;
        self.x = x;
    
    
    def create_list(self):
        for i in range(self.x):
            self.my_list.append(int(input('Enter item: ')));
        return self.my_list;
    
    
    def print_list(self):
        print(' '.join([str(i) for i in self.my_list]));
        
    
    
    def max_heapify(self,my_list,i,s):
        left = lambda i : i * 2;
        right = lambda i : (i * 2) + 1;
        l = left(i);
        r = right(i);
        if l <= s and my_list[l - 1] > my_list[i - 1]:
            largest = l;
        else:
            largest = i;
        if r <= s and my_list[r - 1] > my_list[largest - 1]:
            largest = r;
        if largest != i:
            my_list[largest - 1],my_list[i - 1] = my_list[i - 1],my_list[largest - 1];
            self.max_heapify(my_list,largest,s);
    
        
    def build_max_heap(self,my_list):
        #tree height
        
        for i in range(len(my_list) // 2,0,-1):
            self.max_heapify(my_list,i,len(my_list));
            
            
    def heapsort(self):
        self.build_max_heap(self.my_list);
        #print heap
        print('max or min heap is: ');
        self.print_list();
        #sort the max or min heap(here max)
        size = len(self.my_list);
        for i in range(size,1,-1):
            self.my_list[0],self.my_list[size - 1] = self.my_list[size - 1],self.my_list[0];
            size -= 1;
            self.max_heapify(self.my_list, 1, size);

my_list=[];            
h = HeapSort(my_list, int(input('Enter size of list: ')));       
my_list = h.create_list();
print('before sorting: ');
h.print_list();
h.heapsort();
print('after sorting: ');
h.print_list();    
    